build-support/bin/cherry_pick/helper.js

```diff
@@ -0,0 +1,164 @@
+class CherryPickHelper {
+  constructor({ octokit, context, core }) {
+    this.octokit = octokit;
+    this.context = context;
+    this.core = core;
+    this.pull_number =
+      this.context.issue.number || this.context.payload.inputs.PR_number;
+  }
+
+  get #issue_context() {
+    return {
+      ...this.context.repo,
+      // NB: In GitHub nomenclature, pull requests are (also) issues.
+      issue_number: this.pull_number,
+    };
+  }
+
+  get #pull_context() {
+    return {
+      ...this.context.repo,
+      pull_number: this.pull_number,
+    };
+  }
+
+  get #run_link() {
+    return `:robot: [Beep Boop here's my run link](${this.context.serverUrl}/${this.context.repo.owner}/${this.context.repo.repo}/actions/runs/${this.context.runId}/jobs/${this.context.job})`;
+  }
+
+  #get_branch_name(milestone_title) {
+    return `cherry-pick-${this.pull_number}-to-${milestone_title}`;
+  }
+
+  async #add_failed_label() {
+    await this.octokit.rest.issues.addLabels({
+      ...this.#issue_context,
+      labels: ["auto-cherry-picking-failed"],
+    });
+  }
+
+  async #add_comment(body) {
+    await this.octokit.rest.issues.createComment({
+      ...this.#issue_context,
+      body,
+    });
+  }
+
+  async #get_relevant_milestones(milestone_title) {
+    const semver = require("semver");
+    const { data: all_milestones } =
+      await this.octokit.rest.issues.listMilestones({
+        ...this.#issue_context,
+        state: "open",
+      });
+    const sorted_milestones = all_milestones
+      .map((milestone) => milestone.title)
+      .filter((title) => title.startsWith("2.") && title.endsWith(".x"))
+      .sort((a, b) => semver.compare(semver.coerce(a), semver.coerce(b)));
+    const relevant_milestones = sorted_milestones.slice(
+      sorted_milestones.indexOf(milestone_title)
+    );
+    return relevant_milestones;
+  }
+
+  async get_prereqs() {
+    const { data: pull } = await this.octokit.rest.pulls.get({
+      ...this.#pull_context,
+    });
+
+    if (pull.milestone == null) {
+      const { data: workflows } =
+        // NB: GitHub doesn't give you the URL to the workflow, or the filename, when running a workflow.
+        await this.octokit.rest.actions.listRepoWorkflows({
+          ...this.context.repo,
+        });
+      const workflow_url = workflows.workflows.filter(
+        (workflow) => workflow.name === this.context.workflow
+      )[0].url;
+
+      await this.#add_failed_label();
+      await this.#add_comment(
+        `I was unable to cherry-pick this PR; the milestone seems to be missing.
+
+Please add the milestone to the PR and re-run the [Auto Cherry-Picker job](${workflow_url}) using the "Run workflow" button.
```

Comment:
It might be good to phrase actions like this as directed to a specific person, e.g. for this one, the PR merger. In particular, if the PR author isn't a contributor, they won't have these powers, but they may read this as directed to them.

```suggestion
Person who merged: Please add the milestone to the PR and re-run the [Auto Cherry-Picker job](${workflow_url}) using the "Run workflow" button.
```

(potentially could even pull that person out and @ them)

Reply:
Yeah.

FWIW my second act was going to be a new GHA check (extending the label check) to check for a milestone if there's a CherryPick label

Reply:
> FWIW my second act was going to be a new GHA check (extending the label check) to check for a milestone if there's a CherryPick label

👍 

Even with that initial checker (which sounds good), I think it'd be good to have the script still handle the no milestone case. In particular, it might be easy for the script to 'succeed' at doing nothing, if a PR ended up with no milestone. This might mean it clears out the "needs-cherrypick" label without actually doing any cherry picking.

For instance, this sequence seems _plausible_ (maybe not with the current code, but maybe with a small tweak in future):

- someone adds a milestone before CI runs and it passes, but it's removed by the time the PR merged (either accidentally or on purpose)
- cherry picker runs, and ends up with an empty set of milestones
- the "do the appropriate PRs exist" check is trivially true (there's no milestones, so having no PRs is correct ✅ ), and so the script falls through to removing the label